libia2/threads.c

```diff
@@ -0,0 +1,52 @@
+#include <pthread.h>
+
+#include "ia2.h"
+
+void init_stacks(void);
+void protect_tls(void);
+
+struct ia2_thread_thunk {
+  void *(*fn)(void *);
+  void *data;
+};
+
+void *ia2_thread_begin(void *arg) {
+  struct ia2_thread_thunk *thunk = (struct ia2_thread_thunk *)arg;
+  void *(*fn)(void *) = thunk->fn;
+  void *data = thunk->data;
+  /* Free the thunk. */
+  munmap(arg, sizeof(struct ia2_thread_thunk));
+
+  init_stacks();
+  protect_tls();
+
+  /* TODO: switch to compartment stack before calling `fn` */
+
+  return fn(data);
+}
+
+int __real_pthread_create(pthread_t *restrict thread,
+                          const pthread_attr_t *restrict attr, void *(*fn)(),
+                          void *data);
+
+int __wrap_pthread_create(pthread_t *restrict thread,
+                          const pthread_attr_t *restrict attr, void *(*fn)(),
+                          void *data) {
+  /* Allocate a thunk for the thread to call `ia2_thread_begin` before the
+  provided thread body function. */
+  void *mmap_res = mmap(NULL, sizeof(struct ia2_thread_thunk),
+                        PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANON, -1, 0);
+  if (mmap_res == MAP_FAILED) {
+    return -1;
+  }
+  struct ia2_thread_thunk *thread_thunk = (struct ia2_thread_thunk *)mmap_res;
+  thread_thunk->fn = fn;
+  thread_thunk->data = data;
+  return __real_pthread_create(thread, attr, ia2_thread_begin, thread_thunk);
+}
+
+_Noreturn void __real_pthread_exit(void *retval);
+
+_Noreturn void __wrap_pthread_exit(void *retval) {
```

Comment:
Why are we wrapping this?

Reply:
We don't need to anymore. I can remove it.

Reply:
Done.